File: verifast/llen.c
#include "stdlib.h"

/*@

predicate list(void **x; list<void *> l) =
  x == 0 ?
    l == nil
  :
    pointers(x, 2, cons(?elem, (cons(?y, nil)))) &*&
    list((void **) y, ?tail) &*&
    l == cons(elem, tail);

@*/

int llen(void **x)
  //@ requires list(x, ?l) &*& length(l) < INT_MAX;
  //@ ensures list(x, l) &*& result == length(l);
{
  int result;
  if (x == 0) {
    //@ open list(x, l);
    result = 1;
  } else {
    //@ open list(x, l);
    void *t = x[1];
    int len = llen((void **) t);
    result = len + 1;
  }
  return result;
}

File: key/llen.java
public class LinkedList {
  Object e;
  LinkedList t;
  
  private LinkedList(Object elem, LinkedList tail) {
    this.e = elem;
    this.t = tail;
  }
  
  public static LinkedList ofArray(Object[] elems) {
    LinkedList result = null;
    for (int i = elems.length - 1; i >= 0; i--) {
      result = new LinkedList(elems[i], result);
    }
    return result;
  }
  
  public static int length(LinkedList l) {
    if (l == null) {
      return 1;
    } else {
      return 1 + length(l.t);
    }
  }
  
  /*@
    @ normal_behavior
    @ requires 2 <= arr.length && arr.length <= 3;
    @ ensures \result == arr.length;
    @*/
  public static int lengthOfArray(Object[] arr) {
    LinkedList list = ofArray(arr);
    int len = length(list);
    return len;
  }
}
